relay-server/src/metrics_extraction/transactions/mod.rs

```diff
@@ -375,6 +380,79 @@ fn extract_transaction_metrics_inner(
     Ok(())
 }
 
+/// Extracts metrics from the spans of the given transaction, and sets common
+/// tags for all the metrics and spans. If a span already contains a tag
+/// extracted for a metric, the tag value is overwritten.
+fn extract_span_metrics(
+    aggregator_config: &AggregatorConfig,
+    event: &mut Event,
+    metrics: &mut Vec<Metric>, // output parameter
+) -> Result<(), ExtractMetricsError> {
+    // TODO(iker): measure the performance of this whole method
+
+    if event.ty.value() != Some(&EventType::Transaction) {
+        return Ok(());
+    }
+    let (Some(_), Some(&end)) = (event.start_timestamp.value(), event.timestamp.value()) else {
+        relay_log::debug!("failed to extract the start and the end timestamps from the event");
+        return Err(ExtractMetricsError::MissingTimestamp);
+    };
+
+    let Some(timestamp) = UnixTimestamp::from_datetime(end.into_inner()) else {
+        relay_log::debug!("event timestamp is not a valid unix timestamp");
+        return Err(ExtractMetricsError::InvalidTimestamp);
+    };
+
+    // Validate the transaction event against the metrics timestamp limits. If the metric is too
+    // old or too new, we cannot extract the metric and also need to drop the transaction event
+    // for consistency between metrics and events.
+    if !aggregator_config.timestamp_range().contains(&timestamp) {
+        relay_log::debug!("event timestamp is out of the valid range for metrics");
+        return Err(ExtractMetricsError::InvalidTimestamp);
+    }
+
+    if let Some(spans) = event.spans.value_mut() {
+        // Collect the shared tags for all the metrics and spans on this transaction
+        let mut shared_tags = BTreeMap::new();
+
+        if let Some(transaction_name) = event.transaction.value() {
+            shared_tags.insert("transaction".to_owned(), transaction_name.to_owned());
+        }
+        // TODO(iker): must add shared tags: environment, op, release
+
+        for annotated_span in spans {
+            if let Some(span) = annotated_span.value_mut() {
+                // TODO(iker): add span-specific tags here
+
+                if let Some(user) = event.user.value() {
+                    if let Some(value) = get_eventuser_tag(user) {
+                        metrics.push(Metric::new_mri(
+                            MetricNamespace::Spans,
+                            "user",
+                            MetricUnit::None,
+                            MetricValue::set_from_str(&value),
+                            timestamp,
+                            shared_tags.clone(),
+                        ));
+                    }
+                }
+
+                // Add to the span all the tags that exist in the metrics
+                let span_tags = span.tags.get_or_insert_with(BTreeMap::new);
```

Comment:
Best check this with @mitsuhiko, but I feel it would be safer to put a new data bag called `metrics_tags` or similar on the span. I don't know what `Span::tags` is currently used for, but using a different data bag would help prevent any potential conflicts.